From 6d4814004015906b88156c02053536e7c93fffd6 Mon Sep 17 00:00:00 2001
From: Dong Aisheng <aisheng.dong@nxp.com>
Date: Mon, 27 Nov 2017 17:33:47 +0800
Subject: [PATCH 2957/5242] MLK-17074-2 PM / Domains: choose the deepest state
 to enter if no devices using it

commit  7063b2a5548326cf57706c0855d67ece3622f666 from
https://source.codeaurora.org/external/imx/linux-imx.git

For a domain belongs to no devices anymore, let's choose the deepest state
to enter to save power.

Reviewed-by: Frank Li <frank.li@nxp.com>
Reviewed-by: Ranjani Vaidyanathan <Ranjani.Vaidyanathan@nxp.com>
Signed-off-by: Dong Aisheng <aisheng.dong@nxp.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/base/power/domain.c |    4 ++++
 1 file changed, 4 insertions(+)

diff --git a/drivers/base/power/domain.c b/drivers/base/power/domain.c
index 492a5ed..d5b9705 100644
--- a/drivers/base/power/domain.c
+++ b/drivers/base/power/domain.c
@@ -479,6 +479,10 @@ static int genpd_power_off(struct generic_pm_domain *genpd, bool one_dev_on,
 		if (atomic_read(&genpd->sd_count) > 0)
 			return -EBUSY;
 
+		if (!genpd->device_count)
+			/* Choose the deepest state if no devices using this domain */
+			genpd->state_idx = genpd->state_count - 1;
+
 		/*
 		 * If sd_count > 0 at this point, one of the subdomains hasn't
 		 * managed to call genpd_power_on() for the master yet after
-- 
1.7.9.5

